### PR TITLE
[SPARK-23045][ML][SparkR] Update RFormula to use OneHotEncoderEstimator.

### PR DESCRIPTION
**Pull request by MrBago**
## What changes were proposed in this pull request?

RFormula should use VectorSizeHint & OneHotEncoderEstimator in its pipeline to avoid using the deprecated OneHotEncoder & to ensure the model produced can be used in streaming.

## How was this patch tested?

Unit tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by felixcheung** [comment]
we need to get it to kick off R tests - could you touch one of the files under R/?
also please update PR to include [SPARKR]

**Review by felixcheung — CHANGES_REQUESTED** [review]
we need to get it to kick off R tests - could you touch one of the files under R/?
also please update PR to include `[SPARKR]`

**Comment by viirya** [comment]
Actually I think this can be separated in two different PRs. One for OneHotEncoderEstimator and one for VectorSizeHint.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala [hunk 1]
```diff
@@ -228,22 +229,33 @@ class RFormula @Since("1.5.0") (@Since("1.5.0") override val uid: String)
...
       case Seq(term) =>
-        term
+        dataset.schema(term).dataType match {
+          case _: StringType =>
+            val encodedCol = tmpColumn("onehot")
+            // Formula w/o intercept, one of the categories in the first category feature is
+            // being used as reference category, we will not drop any category for that feature.
+            if (!hasIntercept && !keepReferenceCategory) {
+              keepReferenceCategory = true
+              encoderStages += new OneHotEncoderEstimator(uid)
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: As `OneHotEncoderEstimator` can handle multi-column now, can't we just have one `OneHotEncoderEstimator` to fit for all terms at once?

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: Oh, I see. There is one `OneHotEncoderEstimator` for `dropLast=false`.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala [hunk 2]
```diff
@@ -199,6 +199,7 @@ class RFormula @Since("1.5.0") (@Since("1.5.0") override val uid: String)
     val parsedFormula = RFormulaParser.parse($(formula))
     val resolvedFormula = parsedFormula.resolve(dataset.schema)
     val encoderStages = ArrayBuffer[PipelineStage]()
+    val oneHotEncodeStages = ArrayBuffer[(String, String)]()
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: `oneHotEncodeColumns`

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala [hunk 3]
```diff
@@ -230,16 +231,17 @@ class RFormula @Since("1.5.0") (@Since("1.5.0") override val uid: String)
...
-          .setInputCol(indexed(term))
-          .setOutputCol(encodedCol)
         // Formula w/o intercept, one of the categories in the first category feature is
         // being used as reference category, we will not drop any category for that feature.
         if (!hasIntercept && !keepReferenceCategory) {
-          encoder = encoder.setDropLast(false)
+          encoderStages += new OneHotEncoderEstimator(uid)
+            .setInputCols(Array(indexed(term)))
+            .setOutputCols(Array(encodedCol))
+            .setDropLast(false)
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: Here can optimize. You can merge this multiple (probable) OHEs into one. like:
define:
```
val oneHotEncodeColumnsNotDropLast = ArrayBuffer[(String, String)]()
```
and:
```
if (!hasIntercept && !keepReferenceCategory) {
    oneHotEncodeColumnsNotDropLast += indexed(term) -> encodedCol
} else {
   oneHotEncodeColumns += indexed(term) -> encodedCol
}
```

Comment (MrBago) [reply mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: There is at most 1 encoder with `dropLast(false)`, the next line sets `keepReferenceCategory = true` to ensure we won't take this code path for the remaining columns.

**Review by WeichenXu123 — COMMENTED** [review]
LGTM. thanks!

File: R/pkg/R/mllib_utils.R [hunk 4]
```diff
@@ -130,3 +130,4 @@ read.ml <- function(path) {
     stop("Unsupported model: ", jobj)
   }
 }
+
```
Comment (viirya) [inline R/pkg/R/mllib_utils.R]: Once we make sure R tests are passed, I think we can revert this change back.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala [hunk 5]
```diff
@@ -25,7 +25,7 @@ import org.apache.hadoop.fs.Path
 import org.apache.spark.annotation.{Experimental, Since}
 import org.apache.spark.ml.{Estimator, Model, Pipeline, PipelineModel, PipelineStage, Transformer}
 import org.apache.spark.ml.attribute.AttributeGroup
-import org.apache.spark.ml.linalg.VectorUDT
+import org.apache.spark.ml.linalg.{Vector, VectorUDT}
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: `Vector` is unused import here.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaSuite.scala [hunk 6]
```diff
@@ -17,18 +17,29 @@
 
 package org.apache.spark.ml.feature
 
-import org.apache.spark.{SparkException, SparkFunSuite}
+import org.apache.spark.SparkException
 import org.apache.spark.ml.attribute._
-import org.apache.spark.ml.linalg.Vectors
+import org.apache.spark.ml.linalg.{Vector, Vectors}
```
Comment (viirya) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaSuite.scala]: ditto.

**Review by viirya — APPROVED** [review]
LGTM with minor comments.

**Review by felixcheung — COMMENTED** [review]
R tests passed

**Comment by MrBago** [comment]
I've rebased on master, I think that should resolve the the issues @viirya raised.

**Review by felixcheung — APPROVED** [review]

**Review by viirya — APPROVED** [review]
LGTM

**Review by WeichenXu123 — APPROVED** [review]

**Comment by jkbradley** [comment]
This should go into branch-2.3 to fix the deprecation warning from the old OneHotEncoder, so I'll merge it with master and branch-2.3.  Thanks @MrBago  and everyone who reviewed!
